pip_audit/_dependency_source/requirement.py

```diff
@@ -86,81 +83,45 @@ def collect(self) -> Iterator[Dependency]:
 
         Raises a `RequirementSourceError` on any errors.
         """
-        collected: dict[str, Dependency] = dict()
+
+        # Figure out whether we have a fully resolved set of dependencies.
+        reqs: list[InstallRequirement] = []
+        require_hashes: bool = self._require_hashes
         for filename in self._filenames:
-            try:
-                rf = RequirementsFile.from_file(filename)
-                if len(rf.invalid_lines) > 0:
-                    invalid = rf.invalid_lines[0]
-                    raise InvalidRequirementSpecifier(
-                        f"requirement file {filename} contains invalid specifier at "
-                        f"line {invalid.line_number}: {invalid.error_message}"
-                    )
+            rf = RequirementsFile.from_file(filename)
+            if len(rf.invalid_lines) > 0:
+                invalid = rf.invalid_lines[0]
+                raise InvalidRequirementSpecifier(
+                    f"requirement file {filename} contains invalid specifier at "
+                    f"line {invalid.line_number}: {invalid.error_message}"
+                )
+
+            # If one or more requirements have a hash, this implies `--require-hashes`.
+            require_hashes = require_hashes or any(req.hash_options for req in rf.requirements)
+            reqs.extend(rf.requirements)
+
+        # If the user has supplied `--no-deps` or there are hashed requirements, we should assume
```

Comment:
This section needs to be reviewed carefully.

One weird side effect of this approach is that by avoiding `pip`, we can get different results when auditing fully resolved requirements files versus invoking `pip` since there are now 2 completely separate modes of operation. Our code to handle pre-resolved dependencies has to skip certain types of requirements (URL requirements and editable) whereas the unresolved dependency flow where we invoke `pip` will handle these without a problem. So it's possible that just by adding hashes to a requirements file, users may get a less comprehensive audit.

I think there's a discussion to be had about whether it makes sense to keep this "preresolved" optimisation since `pip install` could be more efficient than our previous homegrown resolver. The new method won't have the problem we used to have where we potentially install a sdist for each candidate just to figure out its dependencies.

Reply:
Thanks for highlighting this! I agree that this isn't ideal -- `--require-hashes` should provide *more* consistency and accuracy during dependency audits, not less. 

> I think there's a future discussion to be had about whether it makes sense to keep this "preresolved" optimisation since pip install could be more efficient than our previous homegrown resolver. The new method won't have the problem we used to have where we potentially install a sdist for each candidate just to figure out its dependencies.

Yeah, let's open up a discussion issue for this. I'm not opposed to removing this special case if the performance hit isn't as significant as the regression we had with `2.4.15`.